kubeflow/kubeflow-training-operator/k8s/imagenet-mpijob.yaml

```diff
@@ -0,0 +1,100 @@
+apiVersion: "kubeflow.org/v2beta1"
+kind: "MPIJob"
+metadata:
+  name: "imagenet-16gpu-mpijob"
+spec:
+  slotsPerWorker: 8
+  runPolicy:
+    cleanPodPolicy: Running
+  mpiReplicaSpecs:
+    Launcher:
+      replicas: 1
+      restartPolicy: OnFailure
+      template:
+        spec:
+          containers:
+            - name: pytorch
+              image: navarrepratt/pytorch_mpi_resnet50:1
+              command:
+                - "mpirun"
+                - "-np"
+                - "16"  # Total processes = num workers * slots per workers
+                - "-x"
+                - "WANDB_API_KEY=$(WANDB_API_KEY)"
+                - "--allow-run-as-root"
```

Comment:
TLDR: I can't get good distributed performance with `MPIJob`.

To preface the below charts, the samples per second number come from `batch size / seconds spent on batch`. The numbers for 8/16 GPUs is an average across all the GPUs. Therefore if the 1 and 8/16 GPU samples per sec were the same, we'd have a perfect scaling factor.

These ran on A40 nodes with the following resource requests:
```yaml
resources:
  requests:
    cpu: 16
    memory: 128G
    nvidia.com/gpu: 8
  limits:
    cpu: 16
    memory: 196G
    nvidia.com/gpu: 8
```

This first chart shows the `MPIJob` performance on 1, 8, and 16 GPUs and clearly there's a big performance drop off. 
![image](https://user-images.githubusercontent.com/24724407/211619205-f6ee638d-6b43-4d19-9940-3f0c7cd574b5.png)

This next chart shows the `PyTorchJob` performance on 8 and 16 GPU. Much better performance and there's hardly a drop off for per GPU throughput when bumping up to 2 nodes.
![image](https://user-images.githubusercontent.com/24724407/211619719-f4c0738f-9195-44ab-8a4b-e762d15aaa46.png)

I tried also setting the following `mpirun` flags according to [horovod's docs](https://horovod.readthedocs.io/en/stable/mpi_include.html#mpirun) but that didn't impact the performance. Anyone have a good guess to what's going on?
```
-bind-to none -mca pml ob1 -mca btl ^openib -x NCCL_SOCKET_IFNAME=eth0
```

Reply:
What GPU types? You should have about 20-30% comms overhead going to multi GPU-multi node. I am surprised by the big drop-off you are having in the MPI Operator example, and the small one in the PyTorch operator. `NCCL_DEBUG=WARN` will show you how NCCL communicates. The only instance that would cause a big discrepancy is if it does very in-efficient intra-node comms in one instance but not the other. If you get scheduled on the same nodes in both cases that should not be a risk (you probably want to add affinities so you run on the same 2 nodes in each example to make sure it's not a node difference).

Reply:
I was using A40's in LAS1 for both. I'll try running with `NCCL_DEBUG=WARN` and on the same two nodes for each type of job.

Reply:
Add an affinity so you run on exactly the same nodes to weed out any inconsistencies.

Reply:
Also for added context, because of how the PyTorch job has the master do compute, I go the single node (8 GPU) numbers for the second chart by running a normal k8s `Job` on a single 8 GPU container with the same `torchrun` command as you see in the yaml for the `PyTorchJob`.

For the MPIJob numbers I always ran it via `MPIJob` but I would change the number and resources of the worker containers.

Reply:
Update:
With some changes described below, I was able to get the MPI Job's samples/sec to not drop off a cliff when going from 1 to 2 nodes. All of these numbers come from running on nodes `g08fb5e` and `g08fb06` with ~full node resources (90 cpu and 700G mem).

MPIJob:
![image](https://user-images.githubusercontent.com/24724407/211890701-d4e4edb6-81bf-4b9a-88e1-fee9ae8022dd.png)

PyTorchJob:
![image](https://user-images.githubusercontent.com/24724407/211890776-c766c723-f049-4d67-b265-ecfb4da597b5.png)

The change that increased the MPIJob's samples/sec and made it stable when going up to 2 nodes was increasing the resource request of the pods. Before, with 16 CPU and 128G mem, the GPU utilization was ~60-80% with 2 nodes and ~20-40% for 1 node. Bumping up the resources allowed the GPU util to get a stable >95% while training.

I initially decreased the resource requests down to 16 CPU, 64G mem, but needed to drop the data loader workers from 16 to 4 to get it to run with going OOM. That tanked a lot of metrics (samples/sec down from the >400 to 100-200). 

The previous PyTorchJob numbers were able to max out the GPU util which is why I think the 1/2 node numbers were already similar. Increasing the resources for the PyTorchJob just allowed it to go faster.

So the scaling factor going from 1 to 2 nodes is almost perfect. My guess as to why it's not as big of an overhead as the expected 20-30% is ResNet-50 is a much smaller model then we usually use with multi-node training. Lots less to communicate between nodes about. [Horovod's own benchmarks](https://horovod.readthedocs.io/en/stable/benchmarks_include.html#benchmarks) report a 90% scaling efficency up to 512 GPUs for ResNet101, double the size of resnet50, so I think our scaling efficiencies make sense. If we want to continue looking into it I can switch to a larger one of [torchvision's image classification models](https://pytorch.org/vision/stable/models.html#classification).

The main open question I still have is the samples/sec performance being almost double on the PyTorchJob vs MPIJob.

Reply:
I was able to get the performance's of pytorch and horovod to line up after adding a `torch.cuda.synchronize()` before stopping the timer in the training loop. Noticed pytorch was doing that themselves in their [DDP benchmarks](https://github.com/pytorch/pytorch/blob/91920ee6da8b8566d0e4bdb4825bfcb4c4938ad5/benchmarks/distributed/ddp/benchmark.py#L78). I couldn't find any "hard" evidence as to why this is needed but I think it comes down to torch's DDP wrapping the model to handle the distribution and horovod wrapping the optimizer. Leads to slightly different race conditions when doing the quick timings for every training batch. 

Once I added that, all of the performance numbers line up. You can see my WandB charts [here](https://wandb.ai/npratt/resnet50-pytorch-vs-horovod) which has 1/2 nodes for torch/horovod.

Reply:
This slows down the PyTorch job by 2x though to align with MPI

Reply:
The total runtime for pytorch and horovod were actually about the same the entire time (with and without `torch.cuda.synchronize()`, ~12 min for 3 epcohs with 16 GPUs.

The samples/sec numbers looks like it'll take twice as long but really it was a sneaky race condition that doesn't impact total runtime. However, I still included the synchronize in the `wandb` if statement since it's not needed if we didn't care about outputting accurate sampels/sec numbers.